pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+	segmentCount := 0
+
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(simultaneousWorkers + 1)
+
+	g.Go(func() error {
+		reader := b.ToReader()
+		defer reader.Close()
+
+		for {
+			// Read either up to EOF or until we have `segmentSize` bytes
+			buf := new(bytes.Buffer)
+			_, err := io.CopyN(buf, reader, int64(a.segmentSize))
+			isEOF := err == io.EOF
+
+			if err != nil && err != io.EOF {
+				logErrorIfNotCancelledContext(err, "Put: Failed to read next segment to write from client. %d of %s (digest function: %s, size: %d)", segmentCount, digest.GetHashString(), digest.GetDigestFunction().GetEnumValue().String(), digest.GetSizeBytes())
+				return err
+			}
+
+			count := segmentCount
+			g.Go(func() error {
+				return a.tables.content.insertSegment(gCtx, blobID, buf.Bytes(), count, digest.GetHashString())
+			})
+
+			segmentCount++
+
+			if isEOF {
+				break
+			}
+		}
+
+		// TODO: Ensure hashes match at this point
+		// TODO: Delete from the blobs table if hashes do not match
+
+		return nil
+	})
+
+	err := g.Wait()
+	if err != nil {
+		logErrorIfNotCancelledContext(err, "Put: Unable to store %s", digest.String())
+		return 0, err
+	}
+	return segmentCount, nil
+}
+
+func (a *cassandraBlobAccess) FindMissing(ctx context.Context, digests bbdigest.Set) (bbdigest.Set, error) {
+	// The obvious thing to do here is to create a `SELECT ... IN (?, ?)` for each digest, but doing so
+	// prevents token-aware routing of queries. Instead, we issue a single `SELECT` per digest and
+	// aggregate the values that come back.
+
+	builder := bbdigest.NewSetBuilder()
+	var mu sync.Mutex
+
+	g, errCtx := errgroup.WithContext(ctx)
+	// Try not to be a terrible, terrible citizen
+	g.SetLimit(simultaneousWorkers)
+	for _, digest := range digests.Items() {
+		g.Go(func() error {
+			return a.isMissing(errCtx, digest, &builder, &mu)
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		logErrorIfNotCancelledContext(err, "FindMissing: returning early because of error reading data")
+		return bbdigest.EmptySet, err
+	}
+
+	toReturn := builder.Build()
+
+	return toReturn, nil
+}
+
+func (a *cassandraBlobAccess) isMissing(ctx context.Context, digest bbdigest.Digest, builder *bbdigest.SetBuilder, mu *sync.Mutex) error {
+	metadata, err := a.tables.metadata.read(ctx, digest)
+	if err != nil {
+		if errors.Is(err, gocql.ErrNotFound) {
+			mu.Lock()
+			defer mu.Unlock()
+			builder.Add(digest)
+			return nil
+		}
+
+		return err
+	}
+
+	go a.tables.metadata.updateLastAccessTime(context.Background(), digest, metadata.instanceName, metadata.lastAccess, a.lastAccessUpdateInterval, a.lastAccessUpdate)
+
+	return nil
+}
+
+func lastUpdateWorker(jobs chan job) {
+	for j := range jobs {
+		j.task()
+	}
+}
+
+func (a *cassandraBlobAccess) getInstanceName(digest bbdigest.Digest) string {
+	instanceName := digest.GetInstanceName().String()
+	if instanceName == a.universalInstanceName {
```

Comment:
Please remove support for "universal instance names". I don't think there's any need to do this as part of this backend specifically. Can't it be implemented as a separate decorator for BlobAccess?

Reply:
Removed. I'll do a follow up PR to set the instance name to a default value if necessary.